### PR TITLE
[proof of concept/ wip] obsm, varm, layers refactor

### PR DESCRIPTION
**Pull request by ivirshup**
@flying-sheep based on our conversation in https://github.com/theislab/scanpy/issues/562#issuecomment-483235556, I figured I'd give it a try. This branch is an attempt at replacing `BoundRecArray` with a `MutableMapping` subclass.

I think this is beneficial largely from a code simplicity point of view. 

* `recarray`s are poorly documented by numpy, and I personally find more difficult than alternatives
* Everyone knows how to use a `dict`
* This can hold sparse arrays and dataframes, unlike recarrays
  * Meaning @flying-sheep's suggested solution for #136 would work
* This could also potentially hold any n>1 dimensional values.

This is WIP since it's not fully featured, and would definitely require a reworking of tests if implemented. One definitely broken thing is IO, but I don't think this will be difficult to solve. Currently this seems to work with the core of scanpy's api (pretty sure all the commands from to seurat based tutorial for `sc.datasets.pbmc3k` work).

### CODE REVIEW

**Comment by ivirshup** [comment]
This is pretty close to being a real implementation now. All non-`Raw` tests now pass for both anndata and scanpy.

Still to go:

* Tests for sparse matrices, dataframes, etc. in `.obsm`, `.varm`
  * Xarray, zarr?
* Investigate on disk representation
* Think about compatibility
  * Since the on disk representation will likely change, how are changes to forwards compatibility handled?

**Comment by flying-sheep** [comment]
Good one!

I think we should figure out how to call the type “any matrix type we support”:

```py
>>> import numpy; numpy.ndarray.mro()                                                                                                                                                                                                                  
[numpy.ndarray, object]
>>> import scipy.sparse; scipy.sparse.spmatrix.mro()                                                                                                                                                                                                         
[scipy.sparse.base.spmatrix, object]
>>> import zarr; zarr.Array.mro()
[zarr.core.Array, object]
>>> import xarray; xarray.DataArray.mro()                                                                                                                                                                                                              
[xarray.core.dataarray.DataArray,
 xarray.core.common.AbstractArray,
 xarray.core.common.ImplementsArrayReduce,
 xarray.core.common.DataWithCoords,
 xarray.core.arithmetic.SupportsArithmetic,
 xarray.core.common.AttrAccessMixin,
 object]
```

So they don’t share any ancestor.

```py
>>> {c: set(dir(c)) & {'toarray', '__array__', 'shape', 'ndim'} for c in [numpy.ndarray, scipy.sparse.spmatrix, zarr.Array, xarray.DataArray]}
{numpy.ndarray: {'__array__', 'ndim', 'shape'},
 scipy.sparse.base.spmatrix: {'ndim', 'shape', 'toarray'},
 zarr.core.Array: {'__array__', 'ndim', 'shape'},
 xarray.core.dataarray.DataArray: {'__array__', 'ndim', 'shape'}}
```

but they do all have `shape`, `ndim`, and either `__array__` or `toarray`!

- So we could accept anything that has that.
- When slicing the whole thing and that fails, we need to emit a good error message telling people which obs exact entry in which annotation failed to be slicable.

**Comment by ivirshup** [comment]
I was thinking I'd go for more of a duck typing approach for now, with types that we explicitly support covered by tests. A large part of that is because I'd like to reduce the total amount of code in AnnData (especially `base.py`!), apart from tests. I'd be up for introducing that kind of abstract type once I see how it will reduce complexity/lines of code.

Question about `Raw`, think that we might remove it in favor of `layers`? I'm trying to figure out how much effort I should put into making it work (i.e., workaround vs refactor).

**Comment by flying-sheep** [comment]
ABCs are duck typing.

**Comment by falexwolf** [comment]
Looks really interesting: `Raw` has a different functional role (different number of variables, only row slicing) than `layers` (same dimensions as `.X`, row and colum slicing). 

Thoughts on treating `connectivities` etc. within that instead of in `.uns`? It's a big flaw of views on  `.uns` that one needs a deepcopy. And this is just due to slicing of the `n_obs` x `n_obs` graph adjacency matrices.

PS: Brave undertaking for something that is mostly about simplifying the code. But, I, of course, don't mind if you have the bandwidth. :)

**Comment by falexwolf** [comment]
Further comments (hopefully not demotivating, just to make sure we're all on the same page):
* You mean n>2 and not "any n>1 dimensional values", right? `obsm` already holds n>1 dimensional values like `X_pca` etc.
* We made `BoundRecArray` so that it can be used like a dict. It has `.keys`, you can `del .obsm[key]` and you can append etc. So, I'd be surprised if usage became even simpler as it is right now.
* There could be some advantage to having a dedicated slot for dense representations and another one for sparse ones. I'd need to think a little about different arguments.

**Comment by ivirshup** [comment]
@flying-sheep  Just to preface, the aphorism for duck-typing I'm familiar with is "if it looks like a duck, and acts like a duck, it's a duck." I googled to make sure I was remembering it right, but it turns out there's a bunch of ways of saying it.

I think creating an ABC would definitely cover "looks like a duck", but I'm not sure if covers "acts like a duck". It lets us know it has those methods, but not necessarily that they result in what we want. If the types were actually subclassed from common ABC I'd be more comfortable with having expectations about their behaviour, like how we expect `.update` and `.items` to work for a mutable mapping. For now I think it's easier to say both "it should have shape", and "when I index into it, it'll return something of the shape I expect" via runtime checks.

I also don't want to get overly ambitious with this. I definitely want this to support sparse arrays, ndarrays, and dataframes. I'll worry about the others once that's accomplished. Could you point me towards the other section you mentioned? I'd be interested in seeing that, as it could give me a better idea of the areas of application.

**Comment by ivirshup** [comment]
@falexwolf Yeah, I think simplifying the code will eventually make it easier to build on top of it. I'd also like to understand how AnnDatas are initialized, and right now it's kinda hard to follow. This seems like a (relatively) simple step towards that. Also simplifying now could reduce additional complexity later, like that this could remove the need for a special attribute for sparse annotations.

I definitely think `connectivities` could fit in `obsm`. That said, `adata.uns["neighbors"]` has a nested structure which is useful for keeping those related values together. I'm not sure what the best way to handle this is.

On your specific points:

* I was thinking that this would allow `obsm` to hold any obs annotations with greater than one dimension. So this includes the current `n==2` case, but I'm not sure the current implementation supports `n==3`.
* What inspired me to do this is that `BoundRecArray` limits the types of values more than I think is necessary for what it does. I think it could be generalized, with easier to follow code. For instance: the current `__setitem__` (https://github.com/theislab/anndata/blob/81d10e2f960873c0283b87ae9f7f983100a8d675/anndata/base.py#L115) is pretty complicated. Also inspired by the pain from #138.
* I'd be interested to hear the advantages. To me, I see `.obsm` as being array-like annotations of the observations, but I'd be interested to hear how you see it.

**Comment by ivirshup** [comment]
@falexwolf and @flying-sheep
More specifically, I've hit a snag with figuring out what to do with setting values on views of an AnnData. I'd like to get your opinions on what the spec for this is.

I've got a little code sample that could be illustrative. I think it could be useful for you to look at it and decide what you think should happen, then try and see if what does happen matches up.

```python
import scanpy as sc
import numpy as np

pbmc = sc.datasets.pbmc68k_reduced()
v = pbmc[pbmc.obs["louvain"] == "1"]
pbmc.obsm["zeros"] = np.zeros_like(pbmc.obsm["X_pca"])
v._isview # True
v.obsm["zeros"] # What should this be?

frozen_obsm = v.obsm.copy()

v.obsm["ones"] = np.ones_like(v.obsm["ones"])
v._isview # Should this be a view?

v = pbmc[pbmc.obs["louvain"] == "1"]
v # now has "zeros" field
v.obsm = frozen_obsm # What should this do? Is v still a view? If so, what's in `obsm`?
v.obsm = v.obsm # What about this?
```

**Comment by falexwolf** [comment]
> I think simplifying the code will eventually make it easier to build on top of it.
Yes, you're absolutely right!

Everything else above also makes a ton of sense! I'm aware that `anndata/base.py` is way too long and was developed way too iteratively just to satisfy certain feature requirements. It can be _much_ simplified and better structured.

> setting values on views 

You shouldn't be able to set any value on views, except for writing to the data matrix. This is what is crucially necessary in the backed case (where you see a portion of the huge data on disk in a view and want to be able to modify it) and adds a lot of convenience in the memory-case (otherwise one wouldn't be able to write into the data matrix using indexing based on obs and var names...

But all other write operations are meaningless (I think), and that's why `DictView`, `ArrayView` etc. exist. They listen to whether you are modifying them and auto-generate a copy if you do so. This is also inspired by the "auto-view-switch" in numpy arrays.

```
v.obsm["ones"] = np.ones_like(v.obsm["ones"])
v._isview # Should this be a view?
```

This is should not be a view anymore (btw, there should be a public `v.isview`). Should also be in tests; if it's a view, it's a bug. I think this should also answer your other questions.

----

> I definitely think connectivities could fit in obsm. That said, adata.uns["neighbors"] has a nested structure which is useful for keeping those related values together. I'm not sure what the best way to handle this is.

Cool! There is a special case about `n_obs` x `n_obs` sparse matrices though, as they need to be both row- and column-sliced, which is evidently not the default for fields in `.obsm`. Don't worry about a nested structure, we don't need it. The idea is to continue to have each tool 'x' write a dict of annotations to `.uns['x']` as long as these things are nothing that we would want to slice. The only case where this occurs is `sc.pp.neighbors` with annotations `distances` and `connectivities`; `params` would stay in `.uns['neighbors']`, for instance.

**Comment by ivirshup** [comment]
### Setting values on views

> backed case

I hadn't been thinking of this. I'll have to look into it for a bit, but will mostly talking about the in-memory case in this comment.

> You shouldn't be able to set any value on views, except for writing to the data matrix.

It's good to have this formalized. The current implementation of the `.obsm` property looks like it has code to update the parent if it's a view, though it doesn't seem to do this in practice.

I do think there could be API benefits to allowing modification via views. Some examples:

```python
# Run a pca just on expression data
sc.pp.pca(adata[:, adata.var["modality"] == "rna"])

# Differential expression, but only within one batch
sc.tl.rank_genes_groups(adata[adata.obs["batch"] = "1"], groupby="celltype")
```

Another case is raised by theislab/scanpy#612 and is relates to 

> "auto-view-switch" in numpy arrays

If the model is numpy arrays, should `adata[subset, :].{some subsetable attribute} = {values}` work? This could be nice for when you do want to modify some array inplace, but want the indexing semantics of an AnnData.

As a side note, I personally don't like the automatic switching in numpy arrays. I think it should be explicit whether an operation returns a view or a copy but for numpy arrays the following code (without type checking) is ambiguous: `b = a[idx]`.

### Subsetting `connectivities`

That's a good point. Subsetting a graph can be pretty different from subsetting a matrix, in the sense that you can subset a graph on edges or nodes.

**Comment by flying-sheep** [comment]
> But all other write operations are meaningless (I think), and that's why DictView, ArrayView etc. exist. They listen to whether you are modifying them and auto-generate a copy if you do so.

FYI: that’s called a “copy on write” structure, or so I think `COWDict` would be a more fitting name here.

**Comment by ivirshup** [comment]
@flying-sheep by "this section" I meant whatever you were referring to by this:

> we’re using a check like that already somewhere else

**Comment by flying-sheep** [comment]
Mostly regular instance checks for different things we support:

https://github.com/theislab/anndata/blob/81d10e2f960873c0283b87ae9f7f983100a8d675/anndata/base.py#L51-L61

https://github.com/theislab/anndata/blob/81d10e2f960873c0283b87ae9f7f983100a8d675/anndata/base.py#L831-L838

https://github.com/theislab/anndata/blob/81d10e2f960873c0283b87ae9f7f983100a8d675/anndata/base.py#L743-L758

**Comment by ivirshup** [comment]
Just removed `BoundRecArray`s entirely from the codebase. Probably still have some cleaning up to do however. Next on my list is adding some tests to make sure views are working, and getting `.raw` to work. I also want to come up with a better name that `DictM`, since that was when I wasn't sure if I'd try to merge this. I'm thinking something like: `AlignedAnnot`?

## Questions

* There was one case I wasn't clear on for views:

```python
pbmc = sc.datasets.pbmc68k_reduced()
v = pbmc[pbmc.obs["louvain"] == "1"]
pbmc.obsm["zeros"] = np.zeros_like(pbmc.obsm["X_pca"])
v._isview # True
v.obsm["zeros"] #  What should this return? Currently is a KeyError
```

* Do you have opinions on changing the on disk representation? I think making `obsm` and `varm` groups instead of datasets will be necessary for supporting more container types, particularly sparse arrays.

**Comment by ivirshup** [comment]
Update on on-disk representation:

The reader and writer handle `obsm`/`varm` being a mapping (encoding it as a group) totally fine. Additionally sparse matrices seem to be handled with current implementation. Dataframes are being read wrong, however.

**Comment by ivirshup** [comment]
@flying-sheep I've been thinking about this bit, and looking over the code for `AnnDataLayers` and thinking this could be simplified even more. I'm thinking these could easily share a parent class (likely an ABC) since they're very similar concepts, and I think handling the "which axes to subset by" can be dealt with by passing subsetting indices as a tuple. Plus it'd be nice if the APIs were the similar, and this would help enforce that.

I've got a skeleton of the class implementations [here](https://gist.github.com/ivirshup/1d50dcc38b1803e1a60e202d784329f4). Mind taking a look?

**Comment by flying-sheep** [comment]
Looks great! I just have design considerations (many of which are actually about the current design):

1. COW, not View. Views propagate changes made to them to the original. Since we have the behavior that we create a copy once we change something, we don’t have views but “copy on write” (aka COW) structures.
2. Use `None` instead of `'NoneType'`. It’s a bit confusing that None gets used as a synonym for it’s own type, but they decided on that
3. I don’t understand what `idx` is for, can you explain? Will it just hold the COW’s subset into the actual structure? Then it should be part of the `AlignedCOWMixin`, right? You should also should fill in the type parameters for `Tuple[?]`. `Tuple[str, ...]` [sic] means that it’s a tuple of strings of unspecified length.
4. Remove `as_dict`. `dict(thing)` will automatically do the same as `dict(thing.items())` if `thing` has an `.items()` method (which we have, being a `MutableMapping`)
5. I don’t think we need to override abstract properties as properties. I think we can just do

    ```py
    class LayersBase(AlignedMapping):
        attrname = "layers"
    ````

**Comment by ivirshup** [comment]
1. It'd be good to represent how modification happens in the name, but I think it's also important to make it clear it's a subset of an existing object (I don't think we necessarily want `"View"`s of the whole object). Maybe something like `ImmutableView`?

2. I always forget that. That `type(None)` returns `NoneType` doesn't help.

3. Currently `AnnDataLayers` has `_oidx` and `_vidx` attributes. `self.idx` is basically `(self._oidx, self._vidx)` since indexing into an array like `a[oidx, vidx]` is the same as `a[(oidx, vidx)]`. For obsm, `idx` would be a tuple like `(oidx, slice(None))`.

4. I'm keeping the contents untyped, because you can index with a bunch of different types, and some objects have different indexing semantics (like `xarray.Dataset`). The types could potentially be parametric, but I think it's simpler to keep it generic for now.

5. `as_dict` is only there for backwards compat right now, though I think that can be broken. Another thing that I think could be broken is the attribute access to the layers.

6. Can do, I'd only checked if that'd work (it doesn't) for instance attributes.

**Comment by flying-sheep** [comment]
1. Good point. `COWView`? `ViewCOW`? Then we can leave the `view`/`isview` and other function names as long as we document what we really mean.
2. :+1: 
3. Understood!
4. Very good opportunity to document this vague “bunch of different types” right now. We can use a [`TypeVar`](https://docs.python.org/3/library/typing.html#typing.TypeVar) with all possible index types in it.
5. Yes. Let’s break them and add an explicit error function instead:

    ````py
    def as_dict(self):
        raise AttributeError(f'{type(self).__name__} does not have `as_dict() anymore. Use `dict(obj)`')
    ```

6. :+1:

**Comment by ivirshup** [comment]
@flying-sheep, mind taking another look? I've managed to get everything running on the same code, and having the tests pass

Things left to do:

* Name things so COW is clear
* Maybe `as_dict` should return a warning for now? It breaks a lot of things if I remove it.
* It'd probably be good to add tests making sure copies are being made when expected
* General code cleanup (probably new files, doesn't make sense for everything to be in `layers.py`)

**Comment by ivirshup** [comment]
Thanks for the review @flying-sheep ! Just getting to it now.

Two general design questions:

## New files

I want to reorganize the code from `base.py` and `layers.py`. I'm thinking something like: 

* `base/anndata.py`
  * Contains `AnnData` and `Raw`
* `base/views.py`
  * `ArrayView`, `DictView`, etc.
* `base/alignedarrays.py`
  * `AxisArrays`, `Layers`

## COW views

This has two parts:

* Should `adata[{some_indexing}].obsm[key][{other_indexing}]` return one of the `ArrayView`, `DataFrame`, other subclasses of `_ViewMixin`? This isn't the case for layers currently, but would make sense to me.
* If views from anndata are all COW, could we just document that and keep the simple names? I think we could just make it clear that the view is attached to an `AnnData` in the `__repr__`.
  * The docs could use a user guide, and this would be in a subsetting section.

File: anndata/base.py [hunk 1]
```diff
@@ -877,23 +772,27 @@ def _init_as_actual(
...
         try:
-            if not isinstance(obsm, np.ndarray) and not obsm:
-                obsm = np.empty(self._n_obs, dtype=[])
-            if not isinstance(varm, np.ndarray) and not varm:
-                varm = np.empty(self._n_vars, dtype=[])
+            if obsm is None:
+                obsm = dict()
+            if varm is None:
+                varm = dict()
         except TypeError:
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/base.py]: I believe I kept the try except because some test was failing without it. Looks like that's been fixed.

I switch back and forth on whether I prefer `dict` or `{}`. I currently prefer `dict()` and `list()` to `{}` and `[]` since it's more consistent with other types. Plus `{}` seems ambiguous with set literals.

File: anndata/base.py [hunk 2]
```diff
@@ -504,17 +395,21 @@ def varm(self):
...
+    def obs_names(self):
+        return self._adata.obs_names
+
     def __getitem__(self, index):
         oidx, vidx = self._normalize_indices(index)
         if self._adata is not None or not self._adata.isbacked: X = self._X[oidx, vidx]
         else: X = self._adata.file['raw.X'][oidx, vidx]
-        if isinstance(vidx, (int, np.int64)): vidx = slice(vidx, vidx+1, 1)
+        if isinstance(vidx, (int, np.int64)): 
+            vidx = slice(vidx, vidx+1, 1)  # why this?
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/base.py]: Better yet, I think I can replace it with an explanatory comment. This is a slice instead of an int in order to keep `adata[:, x: int].shape == adata[:, x: int].X.shape`, right?

File: anndata/base.py [hunk 3]
```diff
@@ -1538,7 +1433,7 @@ def transpose(self) -> 'AnnData':
             raise ValueError(
                 'You\'re trying to transpose a view of an `AnnData`, which is currently not implemented. '
                 'Call `.copy()` before transposing.')
-        layers = {k:(v.T.tocsr() if sparse.isspmatrix_csr(v) else v.T) for (k, v) in self.layers.items(copy=False)}
+        layers = {k:(v.T.tocsr() if sparse.isspmatrix_csr(v) else v.T) for (k, v) in self.layers.items()}
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/base.py]: I don't think I wrote this conditional, looks like it happened a while ago: https://github.com/theislab/anndata/commit/f3fc91218db092c0965eb4ca6cc4952eba354035

It does sorta make sense to me though, `sparse.csr_matrix.T` returns a `csc_matrix`. I'm not sure why we would only do this for csr matrices though.

File: anndata/base.py [hunk 4]
```diff
@@ -877,23 +772,27 @@ def _init_as_actual(
...
-        self._varm = BoundRecArr(varm, self, 'varm')
+
+        # TODO: Probably do this much earlier, potentially for backwards compat
+        # Think about consequences of making obsm a group in hdf
+        if isinstance(obsm, np.ndarray):
+            obsm = convert_to_dict(obsm)
+        if isinstance(varm, np.ndarray):
+            varm = convert_to_dict(varm)
+        self._obsm = AxisArrays(self, 0)
+        self._obsm.update(obsm)
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/base.py]: Good catch. I changed this code before I implemented that part of the constructor and forgot to clean it up.

File: anndata/layers.py [hunk 5]
```diff
@@ -1,96 +1,251 @@
...
+    def parent(self) -> "AnnData":
+        return self._parent
+
+    @abstractmethod
+    def copy(self):
+        pass
+
+    @abstractmethod
+    def view(self, parent, idx):
+        """Returns a subetted copy-on-write view of the object"""
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/layers.py]: Fixed typo

File: anndata/layers.py [hunk 6]
```diff
@@ -1,96 +1,251 @@
...
+        return key in self.parent_mapping
+
+    def __iter__(self):
+        return iter(self.parent_mapping)
+
+    def __len__(self):
+        return len(self.parent_mapping)
+
+    def view(self, parent: "AnnData", subset_idx):
+        """Returns a subsetted view of this object"""
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/layers.py]: Doc-string now inherited from parent.

File: anndata/layers.py [hunk 7]
```diff
@@ -1,96 +1,251 @@
...
+    @abstractmethod
+    def copy(self):
+        pass
+
+    @abstractmethod
+    def view(self, parent, idx):
+        """Returns a subetted copy-on-write view of the object"""
+        pass
+
+    def as_dict(self) -> dict:  #TODO: Deprecate
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/layers.py]: Used decorator

File: anndata/layers.py [hunk 8]
```diff
@@ -1,96 +1,251 @@
...
+        return dict(self)
+
+
+class AlignedViewMixin:
```
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/layers.py]: Response here: https://github.com/theislab/anndata/pull/140#issuecomment-489480913

**Comment by falexwolf** [comment]
This whole thread is awesome.

Re Renaming: I think the most frequently encountered convention is to call the central submodule `core.py` (or `_core.py`) not `base.py`. Hence, I would do
* `_core.py`
* `_views.py`
* `_alignedarrays.py`

Or the same thing in a directory `_core`. Right now, we get autosuggest for all the modules that aren't actually meant for the user:
image

Re COW views: Both of your suggestions: Agreed! :wink:

File: anndata/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (falexwolf) [reply anndata/base.py]: Yes, that's right!

PS: I think it's `adata[:, x:int]` and not `adata[:, x: int]`. See *slices* under https://www.python.org/dev/peps/pep-0008/#pet-peeves.

[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/base.py]: Cool, I'll fix up that comment.

I'd meant `x: int` to read as a type annotation (`x` is an `int`), but now I'm seeing it's (at best) ambiguous in this context.

[the remark this answers is not in the document]
Comment (falexwolf) [reply anndata/base.py]: Ah, yes, sure... I was already wondering... ;)

File: anndata/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (flying-sheep) [reply anndata/base.py]: I think it’s a matter of taste as the argument goes either way. If you have e.g. `{'item'}` and remove the `'item'` you get `{}` and suddenly have a dict instead of a set. But if you have `set()` and change it to `set('item')` you get `{'i', 't', 'e', 'm'}` which is also not what you wanted.

Fun fact: Guido said that if he could redo Python, he’d use `{:}` as the empty dict and `{}` as the empty set.

File: anndata/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (flying-sheep) [reply anndata/base.py]: Ah, right, got it! “Keep the thing a csr_matrix when transposing”

File: anndata/layers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (flying-sheep) [reply anndata/layers.py]: perfect!

**Comment by ivirshup** [comment]
How about:

```
core/anndata.py
core/views.py
core/alignedarrays.py
```

This would similar to how `pandas`, `xarray`, and `MDAnalysis` do it. I think I prefer non `_private` names for modules containing exported classes.

Also, why does your autocomplete look nicer than mine? I don't get any type info.

**Comment by flying-sheep** [comment]
> If views from anndata are all COW, could we just document that and keep the simple names

I’m always for simple, but never for misleading. “view” has a very specific meaning in python, which is “changes made to this object actually change the object this is a view of”.

When I do

```py
>>> view = obj.view()
>>> view[foo] = bar
>>> obj[foo]
KeyError: ...
```

I’m *very* confused.

**Comment by flying-sheep** [comment]
> Also, why does your autocomplete look nicer than mine? I don't get any type info.

Because he uses a version of IPython where this PR has been merged: ipython/ipython#11506

**Comment by ivirshup** [comment]
I'm not sure it's so cut and dry in the stdlib:

* `memoryview`s  can be mutable
* While there are `MappingViews`, they don't share and api with the parent mapping. This seems like a different definition of view to me – more like currying maybe?
* There is `types.MappingProxyType`, which fits the immutable view definition I think we're agreed on, but it's not called a view.

I'm also not sure we can claim total analogy with numpy. Where `a` is a `np.ndarray`:
 * `a[idx] = val` always modifies in place. `adata[idx] = val` or `adata[idx].attr = val` generally does not.
 * The following is ambiguous. For anndata a similar operation would return `False` unless values were set on `X`, where I believe it should always be `True`

```python
a = np.ones((10, 10))
b = a[idx]
b[:] = 0
np.all(a[idx] == 0)
```

On the `.view` method, I think that seems fair. However that's not how users interact with AnnData, since it doesn't have a `.view` method. This has made me think I should rename the `AlignedMapping.view` method, or at least make it private. Could be `AlignedMapping.subset`?

We could potentially go with something like `Subset` which avoids the more loaded `View` in general? On the other hand, it would be really nice to not break current naming schemes. I think I'm also still pro allowing modifications of views in the future, as it could allow even more data to live on disk in the backed case.

@falexwolf, you mentioned before the inspiration of numpy's  "auto-view-switch" for how views worked. Would you mind elaborating on that/ linking to (possibly external) design docs that describe this? My initial understanding of this was that memory was shared when it's easy, but copies were made when sharing was hard.

**Comment by ivirshup** [comment]
I think I've got that PR merged for mine (I'm on the current version), I've just got jedi disabled until path completion works with it (I assume it's still broken if https://github.com/ipython/ipython/issues/11530 hasn't been closed?)

**Comment by ivirshup** [comment]
Turns out having views of mapping return `_ViewMixin` subclasses becomes more complicated than I expected, since `_ViewMixin` subclasses are assumed to be attributes of an AnnData instead of values of attributes. I think this can be worked around by saying `view_args` is either `Tuple[AnnData, str]` or `Tuple[AnnData, str, str]`. If it's the latter, the second str is used as a key. How does this sound?

This partially solves the deep copy problem for `uns`. Something else I realized (which I think @falexwolf will like), is that having a mapping where values are array-like objects where both axes align to one axis of the parent is a pretty trivial subclass of what I've done already. I think this would fit the behavior you've described previously for cell graphs.

File: anndata/utils.py [hunk 9]
```diff
@@ -1,17 +1,40 @@
...
+        raise TypeError(
+            "Can only convert np.ndarray with compound dtypes to dict, passed "
+            "array had '{}'.".format(obj.dtype)
+        )
+    d = dict()
+    for k in obj.dtype.fields.keys():
+        d[k] = obj[k]
+    return d
+
+@convert_to_dict.register(type(None))  # This is required for python < 3.7
```
Comment (flying-sheep) [inline anndata/utils.py]: You mean the part `(type(None))` is required for those versions, right?

I initially interpreted it as “registering the function for NoneType at all is required for python 3.6” and was like “wait, does `dict(None)`” work in 3.7? I think the comment is more confusing than enlightening :wink:

Comment (ivirshup) [reply anndata/utils.py]: Yeah, but I actually just learned it was required for a different reason than I thought, and `.register(type)` is required for all pre 3.7 versions. Now it's not even necessary I think.

**Comment by ivirshup** [comment]
@flying-sheep, I've been thinking about the renaming, and I think I'd like to keep this PR to non-breaking changes. My goal here is just some backend centralization of code, and adding a few features, but not yet worthy of a breaking version bump.

By splitting out that discussion we can also get more opinions on appropriate naming and behavior, which I think would be valuable. We could also bundle any api changes (like swapping out `isview`) with any other api changes we'd like to make.

Here's what my goals for this PR are:

* Code reorganization, and consolidation of redundant code for mappings with aligned values
* Increased consistency of api, mostly that views have copy-on-write behavior
* Work towards a more flexible AnnData by allowing more types of array-like annotations for obsm, varm, and layers.
  * I think fully implementing this is out of scope for this PR, since I'd like IO changes to be separate.

## What I think is left to do:

- [ ] Return `_SetItemMixin` derived classes from operations like `adata.obsm["x"][idx]` and test they work as expected
- [ ] Add narrative docs specifying that views are copy on write
- [ ] Remove demonstrative `{obs,var}p` attribute

### Kinda related things that could go here

* Fix for #148, since I've already changing code that doesn't fit copy-on-write semantics
* Maybe fix #145, since I'm already touching the code here: https://github.com/theislab/anndata/pull/140#discussion_r281048609

### Issues this could close

* #142
* #136 (since sparse arrays can just go in `{obs,var}m`)
* #45

**Comment by flying-sheep** [comment]
Changing the type names is a non-breaking change, as they’re private types and aren’t documented. Anyone doing `from anndata import ArrayView` now does so at their own risk.

**Comment by ivirshup** [comment]
Oh, I'd thought you'd wanted to rename `.isview` also? Then it could be fine.

Also, isn't `ArrayView` one we wouldn't want to rename? Based on Alex's comments above, don't views have a mutable `.X`?

Either way, it could happen in another PR. I'd also like to play around with making as much of AnnData be backed, and having views on them all be mutable. If that goes well, we might not want to change the name.

**Comment by flying-sheep** [comment]
> Also, isn't ArrayView one we wouldn't want to rename? Based on Alex's comments above, don't views have a mutable .X?

I didn’t think about it, it was just an example :sweat_smile:

**Comment by ivirshup** [comment]
I wanna try. I’ll give a shot at explaining my rationale for this:

If your dataset is really big, it’s not just X that’s gonna use a lot of memory. So are any layers, obsm, varm, etc. If we can already deal with arrays being backed on disk, then it probably shouldn’t matter which array it is. I also think I’m just as likely (if not more) to modify any of the other arrays as I am X.

I also think it'd be convenient if scanpy would do the work of translating indices when I want to modify some data in the object. I think this is what was being tried in #148.

**Comment by ivirshup** [comment]
Just getting back to this, (had my confirmation, got real sick right after), but I think it's getting close to done.

Were we going to go back to 3.5 compatibility for scanpy? If not, are we gonna keep 3.5 compatibility for anndata?

**Comment by falexwolf** [comment]
I'm fine dropping 3.5 compat for AnnData.

**Comment by ivirshup** [comment]
Question about this PR. I'd like to merge it soon, but it will add features that aren't totally supported.

You can now put a dataframe or sparse matrix in obsm, varm, and layers, but those won't survive a roundtrip to disk. What's the best way to handle this until then? I'm thinking setting an unsupported value should either be a warning or an error.

**Comment by falexwolf** [comment]
> I'm thinking setting an unsupported value should either be a warning or an error.

That is a perfect solution! I'm fine if we go ahead with this. Any necessity to release 0.6.21 before?

**Comment by ivirshup** [comment]
Since it's a big code change, I think it'd be nice to have it sit on master for a little bit before a release. Just a chance to catch any bugs I've missed.

I also ended up getting started on the io parts, and have made better progress than I expected. It's possible these can come out together. I'll have to work on it for another day or two to know for sure.

**Comment by falexwolf** [comment]
Great. Completely agree that we shouldn't move quick with the next release once this is on master. Should your other PR (`obs_vector`) go into a release `0.6.21` before or is it not necessary?

**Comment by ivirshup** [comment]
It would be great to have a release of the `obs_vector` stuff first, so I can merge some long standing PRs in scanpy.

I think this is pretty much ready to go, but I think it's going to have to be merged from another PR since there's only so much time I'm willing to spend on managing a git history.

**Comment by falexwolf** [comment]
Good, I'll make a release as soon as Volker has answered some trivial questions around `read_loom(..., cleanup...)`. I'll let you know. Should be later today.

Feel free to merge all of this stuff here in any way you like after `0.6.21` is made. 🙂
